### PR TITLE
HIP-1008: Enhance Mirror Node Contract Log REST API with transaction hash

### PR DESCRIPTION
**Pull request by georgi-l95**
**Description**:
This PR describes the enhancment of the mirror node contract log rest api with transaction hash query parameter

**Related issue(s)**:

Fixes #

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: HIP/hip-1008.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+### Transaction Hash Parameter
+
+A new transaction.hash parameter will be added to search for contracts logs by their hash. This transaction hash query parameter should be in hex format with optional leading `0x` prefix and will support both 32 byte Ethereum and 48 byte HAPI hashes.
+
+**`/api/v1/contracts/results/logs?transaction.hash={value}`**
+
+**Example Request**
+**`/api/v1/contracts/results/logs?transaction.hash=0xda1aee3c386763b2cb2e830765053e346d11deff1c343a5fefe84910f9dfe63b`**
+
+Note: if`limit` is not specified in the request, response will be limited to 25 by default
```
Comment (steven-sheehy) [inline HIP/hip-1008.md:47]:
```suggestion
Note: If `limit` is not specified in the request, the response will be limited to 25 by default.
```

File: HIP/hip-1008.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+The proposed enhancement involves adding query parameter `transaction.hash` to the existing route `/api/v1/contracts/results/logs` to return information about the contract log.
+
+### Transaction Hash Parameter
+
+A new transaction.hash parameter will be added to search for contracts logs by their hash. This transaction hash query parameter should be in hex format with optional leading `0x` prefix and will support both 32 byte Ethereum and 48 byte HAPI hashes.
+
+**`/api/v1/contracts/results/logs?transaction.hash={value}`**
+
+**Example Request**
+**`/api/v1/contracts/results/logs?transaction.hash=0xda1aee3c386763b2cb2e830765053e346d11deff1c343a5fefe84910f9dfe63b`**
```
Comment (steven-sheehy) [inline HIP/hip-1008.md:45]:
```suggestion

`/api/v1/contracts/results/logs?transaction.hash=0xda1aee3c386763b2cb2e830765053e346d11deff1c343a5fefe84910f9dfe63b`
```

File: HIP/hip-1008.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+- As a user of mirror node, I want to be able to query using transaction hash and get the corresponding contract log.
+- As a user of ethereum tools, I want to be able to request transaction receipt using `eth_getTransactionReceipt` and get corresponding HTS operation transaction receipt.
+
+## Specification
+
+The proposed enhancement involves adding query parameter `transaction.hash` to the existing route `/api/v1/contracts/results/logs` to return information about the contract log.
+
+### Transaction Hash Parameter
+
+A new transaction.hash parameter will be added to search for contracts logs by their hash. This transaction hash query parameter should be in hex format with optional leading `0x` prefix and will support both 32 byte Ethereum and 48 byte HAPI hashes.
```
Comment (steven-sheehy) [inline HIP/hip-1008.md:40]:
```suggestion
A new `transaction.hash` parameter will be added to search for contracts logs by their hash. This transaction hash query parameter should be in hex format with optional leading `0x` prefix and will support both 32 byte Ethereum and 48 byte Hedera transaction hashes.
```

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

**Review by steven-sheehy — DISMISSED** [review]
LGTM

**Review by xin-hedera — APPROVED** [review]
LGTM

**Review by Nana-EC — APPROVED** [review]

**Review by Nana-EC — DISMISSED** [review]

**Comment by Neurone** [comment]
Hi guys, how is this different from using the [`/api/v1/contracts/results/{transactionIdOrHash}`](https://mainnet.mirrornode.hedera.com/api/v1/docs/#/contracts/getContractResultByTransactionIdOrHash) endpoint?

For example[, this is just a random mainnet transaction](https://mainnet.mirrornode.hedera.com/api/v1/contracts/results/0xcc7bfea2d35e4bcf27bfe60d19433efa8e03f3d6fbafebe0e21c0f24bcee3d0f), and I can get the logs by querying the mirror node with the 32 byte Ethereum format, and I get all the logs:

```zsh
⮞ curl --silent https://mainnet.mirrornode.hedera.com/api/v1/contracts/results/0xcc7bfea2d35e4bcf27bfe60d19433efa8e03f3d6fbafebe0e21c0f24bcee3d0f | jq
{
  "address": "0x00000000000000000000000000000000003c437a",
  "amount": 2500000000000,
  "bloom": "0x00000080000000102001000000000000000000000000020100000000000020000000000000000000800000000000000000000000010022000000004040020040010040000008200800000000000000000000008000800001000000000000000000000000020000000000000000200800000000000000000000200040000880000000000000000000000000008000000000000000000000000000200000000000000080000004000000000000000000000000000000000000000000000000000000080001000000000000000000000000000000000000000000000000000020000000000000000004000000000800000000000000000080100000000000800000",
  "call_result": "0x0000000000000000000000000000000000000000000000000000000062a349f6",
  "contract_id": "0.0.3949434",
  "created_contract_ids": [],
  "error_message": null,
  "from": "0x000000000000000000000000000000000056dc0c",
  "function_parameters": "0x414bf3890000000000000000000000000000000000000000000000000000000000163b5a000000000000000000000000000000000000000000000000000000000006f89a00000000000000000000000000000000000000000000000000000000000005dc000000000000000000000000756dab6280de64f2c58a2c693a1528fcaf23ec5a0000000000000000000000000000000000000000000000000000000066a18f8900000000000000000000000000000000000000000000000000000246139ca80000000000000000000000000000000000000000000000000000000000629b8d9c0000000000000000000000000000000000000000000000000000000000000000",
  "gas_consumed": 209996,
  "gas_limit": 2000000,
  "gas_used": 1600000,
  "timestamp": "1721863475.979812003",
  "to": "0x00000000000000000000000000000000003c437a",
  "hash": "0xcc7bfea2d35e4bcf27bfe60d19433efa8e03f3d6fbafebe0e21c0f24bcee3d0f",
  "block_hash": "0x702fa733df1414be14a5fa96fa6e487b363fb6e2dfc62fd3b457aa9365e58818673d13d54760641e90e9a11c1c691d1c",
  "block_number": 66915917,
  "logs": [
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "bloom": "0x00000080000000000000000000000000000000000000020000000000000020000000000000000000000000000000000000000000010000000000000000000040000040000000000000000000000000000000008000800000000000000000000000000000000000000000000000000000000000000000000000000000000080000000000000000000000000008000000000000000000000000000200000000000000000000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000000000000000004000000000800000000000000000000000000000000800000",
      "contract_id": "0.0.3964804",
      "data": "0x0000000000000000000000000000000000000000000000000000000062a349f6",
      "index": 0,
      "topics": [
        "0x5f2147fb558c977441fbdfebcf8cd5776606adc8da5ff95566fc2a4137e54d13",
        "0x000000000000000000000000c5b707348da504e9be1bd4e21525459830e7b11d",
        "0x000000000000000000000000756dab6280de64f2c58a2c693a1528fcaf23ec5a",
        "0x000000000000000000000000000000000000000000000000000000000006f89a"
      ]
    },
    {
      "address": "0x0000000000000000000000000000000000163b59",
      "bloom": "0x00000000000000102001000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000004000020000000000000000000000000000000000000000000000000001000000000000000000000000020000000000000000000800000000000000000000000040000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000100000000000000000",
      "contract_id": "0.0.1456985",
      "data": "0x00000000000000000000000000000000000000000000000000000246139ca800",
      "index": 1,
      "topics": [
        "0x831ac82b07fb396dafef0077cea6e002235d88e63f35cbd5df2c065107f1e74a",
        "0x0000000000000000000000000000000000000000000000000000000000000000",
        "0x00000000000000000000000000000000000000000000000000000000003c437a"
      ]
    },
    {
      "address": "0x0000000000000000000000000000000000163b59",
      "bloom": "0x00000000000000102001000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000004000020000010000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000000000000000000080000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000080100000000000000000",
      "contract_id": "0.0.1456985",
      "data": "0x00000000000000000000000000000000000000000000000000000246139ca800",
      "index": 2,
      "topics": [
        "0x831ac82b07fb396dafef0077cea6e002235d88e63f35cbd5df2c065107f1e74a",
        "0x0000000000000000000000000000000000000000000000000000000000163b59",
        "0x00000000000000000000000000000000000000000000000000000000003c437a"
      ]
    },
    {
      "address": "0x0000000000000000000000000000000000163b59",
      "bloom": "0x00000000000000102000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000002000000004040020000000000000008000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
      "contract_id": "0.0.1456985",
      "data": "0x00000000000000000000000000000000000000000000000000000246139ca800",
      "index": 3,
      "topics": [
        "0x5548c837ab068cf56a2c2479df0882a4922fd203edb7517321831d95078c5f62",
        "0x00000000000000000000000000000000000000000000000000000000003c437a",
        "0x00000000000000000000000000000000000000000000000000000000003c437a"
      ]
    },
    {
      "address": "0x00000000000000000000000000000000003c437a",
      "bloom": "0x00000000000000002000000000000000000000000000000100000000000000000000000000000000800000000000000000000000000000000000000000020040000040000000200000000000000000000000008000000000000000000000000000000000000000000000000000200000000000000000000000200000000000000000000000000000000000000000000000000000000000000000200000000000000000000004000000000000000000000000000000000000000000000000000000080001000000000000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000",
      "contract_id": "0.0.3949434",
      "data": "0x00000000000000000000000000000000000000000000000000000246139ca800",
      "index": 4,
      "topics": [
        "0x5f2147fb558c977441fbdfebcf8cd5776606adc8da5ff95566fc2a4137e54d13",
        "0x00000000000000000000000000000000000000000000000000000000003c437a",
        "0x000000000000000000000000c5b707348da504e9be1bd4e21525459830e7b11d",
        "0x0000000000000000000000000000000000000000000000000000000000163b5a"
      ]
    },
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "bloom": "0x00000080000000002000000000000000000000000000020000000000000020000000000000000000800000000000000000000000000020000000000000020000000000000000000800000000000000000000000000800000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000800000000000000000000000000000000800000",
      "contract_id": "0.0.3964804",
      "data": "0xffffffffffffffffffffffffffffffffffffffffffffffffffffffff9d5cb60a00000000000000000000000000000000000000000000000000000246139ca8000000000000000000000000000000000000000026dd1879916d967715f2373c8b00000000000000000000000000000000000000000000000000002c1e4b6246380000000000000000000000000000000000000000000000000000000000011df4",
      "index": 5,
      "topics": [
        "0xc42079f94a6350d7e6235f29174924f928cc2ac818eb64fed8004e115fbcca67",
        "0x00000000000000000000000000000000000000000000000000000000003c437a",
        "0x000000000000000000000000756dab6280de64f2c58a2c693a1528fcaf23ec5a"
      ]
    }
  ],
  "result": "SUCCESS",
  "transaction_index": 2169,
  "state_changes": [
    {
      "address": "0x0000000000000000000000000000000000163b59",
      "contract_id": "0.0.1456985",
      "slot": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "value_read": "0x0000000000000000000000000000000000000000000000000000000000163b5a",
      "value_written": null
    },
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "contract_id": "0.0.3964804",
      "slot": "0x0000000000000000000000000000000000000000000000000000000000000000",
      "value_read": "0x00016603e803e801b8011ddb0000000000000026d0476cd54b31a0432df2b17b",
      "value_written": "0x00016603e803e801b8011df40000000000000026dd1879916d967715f2373c8b"
    },
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "contract_id": "0.0.3964804",
      "slot": "0x0000000000000000000000000000000000000000000000000000000000000001",
      "value_read": "0x00000000000000000000000000000000005fc4cc953ccd3c4fdeea40c6a02008",
      "value_written": null
    },
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "contract_id": "0.0.3964804",
      "slot": "0x00000000000000000000000000000000000000000000000000000000000001c0",
      "value_read": "0x01000033ac000002490666a6d9f3b0cedb28eb8b6b0001621adbef1c66a18d32",
      "value_written": null
    },
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "contract_id": "0.0.3964804",
      "slot": "0x0000000000000000000000000000000000000000000000000000000000000002",
      "value_read": "0x00000000000000000000000000000001827f81feed58a340532bc80beae8f2d2",
      "value_written": "0x0000000000000000000000000000000182839d824daf134d981963efa8a2986f"
    },
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "contract_id": "0.0.3964804",
      "slot": "0x0000000000000000000000000000000000000000000000000000000000000003",
      "value_read": "0x00000000000000000000000d2e7345ae00000000000000000000000001a41f8f",
      "value_written": "0x00000000000000000000000d53b403ee00000000000000000000000001a41f8f"
    },
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "contract_id": "0.0.3964804",
      "slot": "0x0000000000000000000000000000000000000000000000000000000000000004",
      "value_read": "0x00000000000000000000000000000000000000000000000000002da1fa0d32c2",
      "value_written": "0x00000000000000000000000000000000000000000000000000002c1e4b624638"
    },
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "contract_id": "0.0.3964804",
      "slot": "0x1d0d48568343e0c318307d6a3df806b1ed226b712d0eefd7298ef97ba328225d",
      "value_read": "0xfffffffffffffffffffffe7c515513760000000000000000000001edee636f6a",
      "value_written": null
    },
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "contract_id": "0.0.3964804",
      "slot": "0x1d0d48568343e0c318307d6a3df806b1ed226b712d0eefd7298ef97ba328225e",
      "value_read": "0x00000000000000000000000000000000000207b9b3639a188edec81de2be5f18",
      "value_written": "0x00000000000000000000000000000000005dbd12e1d93323c1002222e3e1c0f0"
    },
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "contract_id": "0.0.3964804",
      "slot": "0x1d0d48568343e0c318307d6a3df806b1ed226b712d0eefd7298ef97ba328225f",
      "value_read": "0x000000000000000000000000000000000c00017495c57e889cf0649efc1ad844",
      "value_written": "0x000000000000000000000000000000017682d627f0d0180770cc033e6bd4132a"
    },
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "contract_id": "0.0.3964804",
      "slot": "0x1d0d48568343e0c318307d6a3df806b1ed226b712d0eefd7298ef97ba3282260",
      "value_read": "0x01000d0aee0000000000000015e06a14f2e5c233cea63cd73a00000eaf341118",
      "value_written": "0x0166948244000033ac0000023325fc91e70dee9b0c82aeb4310001536ba7de04"
    },
    {
      "address": "0xc5b707348da504e9be1bd4e21525459830e7b11d",
      "contract_id": "0.0.3964804",
      "slot": "0xca4d0c6c94a9477136dd41a99cc19ecbe441c8f6609efe7c6fa65be007a47323",
      "value_read": "0x00000040810801ff4501ebddfdfdfffffffbfdfefefffffffbffffe7fffffef4",
      "value_written": null
    }
  ],
  "status": "0x1",
  "failed_initcode": null,
  "access_list": "0x",
  "block_gas_used": 3048898,
  "chain_id": "0x127",
  "gas_price": "0x092",
  "max_fee_per_gas": "0x",
  "max_priority_fee_per_gas": "0x",
  "r": "0x57047646cd71876b9c8e3989930de46e259c51b6ca013de841361cca34dbcfff",
  "s": "0x39bb0f720841393039ac885b7fdec31fa4c3efa7d4350c6cb8bea709875618fc",
  "type": 0,
  "v": 626,
  "nonce": 13320
}
```

**Comment by georgi-l95** [comment]
> Hi guys, how is this different from using the [`/api/v1/contracts/results/{transactionIdOrHash}`](https://mainnet.mirrornode.hedera.com/api/v1/docs/#/contracts/getContractResultByTransactionIdOrHash) endpoint?
> 
> For example, [this is just a random mainnet transaction](https://mainnet.mirrornode.hedera.com/api/v1/contracts/results/0xcc7bfea2d35e4bcf27bfe60d19433efa8e03f3d6fbafebe0e21c0f24bcee3d0f), and I can get the logs by querying the mirror node with the 32 byte Ethereum format, and I get all the logs:
> 
> ```shell
> ⮞ curl --silent https://mainnet.mirrornode.hedera.com/api/v1/contracts/results/0xcc7bfea2d35e4bcf27bfe60d19433efa8e03f3d6fbafebe0e21c0f24bcee3d0f | jq
> ```

Hey @Neurone thanks for pointing this out. Difference here is that this endpoint is querying from `contract_results` table in the database, where contract/logs is querying from `contract_logs` table. Not all transactions are logged in the `contract_results` table, for example HTS trasnfer and approve. Their events are only added in the latter database table.

**Review by mgarbs — APPROVED** [review]

**Comment by Neurone** [comment]
Hi Georgi, thanks for the clarification.

I think we should take advantage of the change to match the log values in the `/api/v1/contracts/results/{transactionIdOrHash}` endpoint to the expected results (all logs relative to that transaction).

The DB query should be able to read this information and associate it with the result.

We can also add a `/api/v1/contracts/results/{transactionIdOrHash}/logs` endpoint in case the developers just want to get the logs, and in that case it would be a read from the `contract_logs` table only. This is exactly the behaviour we would have got with the new `/api/v1/contracts/results/logs?transaction.hash={value}` syntax, but now with a more coherent syntax.

This way we both cover the need to have the full log list and address the misunderstanding about data for those projects that use the logs value returned by the `/api/v1/contracts/results/{transactionIdOrHash}`.

If the `contract_logs' table contains all the logs, we can also take the opportunity to free up some space in the `contract_results' table.

@steven-sheehy @xin-hedera What do you think?

**Pull request merged**